doc/source/user/build_system/core_files.rst

```diff
@@ -0,0 +1,257 @@
+* ``depend``: Dependencies on other cores. More on dependencies later.
+
+Source files
+~~~~~~~~~~~~
+
+Source files are resolved relative to the location of the core file and must be stored in the same directory as the core file, or in a subdirectory of it.
+Source file names cannot be absolute paths, or start with ``../``.
+
+Optionally, source files can have attributes; the file ``macros.vh`` is an example of that.
+When specifying attributes, end the file name with a colon (``:``), and specify attributes as key-value pairs below it.
+(Alternatively, the equivalent short form syntax can be used, e.g. ``macros.vh: {is_include_file: true}``.)
+
+The most common attributes are:
+
+* ``is_include_file``: The file is an include file. In Verilog and C/C++, this means the file is not passed to the tool directly, but instead the file is included by another source file.
+  FuseSoC ensures that the tool finds the include file, e.g. by passing an appropriate include path to the tool.
+* ``file_type``: Override the default file type of the fileset for this particular file.
+
+.. todo::
+
+   Link to reference documentation on CAPI2.
+
+File types
+~~~~~~~~~~
+
+A file type describes the type of source file.
+FuseSoC does not use this information itself, but passes it on to tool backends which then configure the tool appropriately depending on the file type encountered.
+
+Commonly used file types are:
+
+* ``verilogSource``: Verilog source code, up to Verilog-2001.
+  Files ending in ``.v`` or ``.vh`` should use this type.
+* ``systemVerilogSource``: SystemVerilog source code (design and test code).
+  Files ending in ``.sv`` or ``.svh`` should use this type.
+* ``vhdlSource``: VHDL source code.
+  Files ending in ``.vhd`` or ``.vhdl`` should use this file type.
+
+.. todo::
+
+   Link to reference documentation on file types.
+
+Targets
+-------
+
+A :term:`target` can be seen as something you would like to do with the source code in the core: synthesize it, simulate it, lint it.
+Targets are specified as dictionaries under the ``targets`` top-level key.
+
+.. literalinclude:: ../../../../tests/userguide/blinky/blinky.core
+   :language: yaml
+   :start-at: targets:
+   :end-at: clk_freq_hz: 100000000
+   :name: ug_build_system_cb_targets
+
+The blinky example shown above defines three targets: the ``default`` target, a ``sim`` target to simulate the design, and a ``synth`` target to synthesize it.
+Many designs also define a ``lint`` target to run static analysis jobs.
+The ``sim`` and ``synth`` targets are optional and could have had any name.
+The ``default`` target is special and required.
+
+Within a target
+~~~~~~~~~~~~~~~
+
+Within each target block multiple keys determine what the target does.
+The most common keys are:
+
+* ``filesets``: An ordered list of file sets (source files) included in the target.
+* ``description`` (optional): A description of the target.
+* ``toplevel`` (optional): The name of the design toplevel.
```

Comment:
Maybe add a note that this can be a list, for designs which require multiple toplevels. Don't want it to be too prominent since it's somewhat rare (although many designs using Xilinx IP need this)

Reply:
Good point. I thought about it and then didn't mention it, but I think we can easily mention it as advanced use. 

I've been thinking about always making it a list (to ease documentation and type hinting), but I'll do so separately once I have finished thinking (which can take a surprisingly long time) :smile:

Reply:
Please don't. It was a list originally and IIRC it was you who wanted to have it as a string. I think the current way is good as it makes the common case easy but still make the special case possible

Reply:
Wait, I'm contradicting myself? Can't be!!!!!11